### PR TITLE
[SYCL][CUDA] Add mem_advise reset and managed mem check

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This PR adds the mem_advise reset case and a managed memory check in cuda_piextUSMEnqueueMemAdvise.

cuMemAdvise only works on managed memory and returns an error if host or device memory is passed.
The SYCL-CTS tests mem_advise with host and device memory, which are not managed.  This PR prevents an error from being thrown in the cts usm test.

In addition, the sycl spec specifies for mem_advise `A value of 0 reverts the advice for ptr to the default behavior`, currently a value of 0 is treated as unknown and throws an error in the cts as well. This reset case is added to mem_advise.

This patch along with #5446 resolves issue #5209 

This change breaks the ABI

### CODE REVIEW

File: sycl/include/CL/sycl/detail/pi.h [hunk 1]
```diff
@@ -448,7 +448,8 @@ typedef enum {
 
 typedef enum {
   // Device-specific value opaque in PI API.
-  PI_MEM_ADVICE_UNKNOWN,
+  PI_MEM_ADVISE_RESET = 0,
+  PI_MEM_ADVICE_UNKNOWN = 1,
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/detail/pi.h:452]: Is this enum visible to the user?
I reckon this change breaks ABI.

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h:452]: We should not change values in the enum, just add new

**Review by s-kanaev — CHANGES_REQUESTED** [review]
Please, don't change user-visible enum values. Only add new ones.

File: sycl/include/CL/sycl/detail/pi.h (hunk 1 above)
Comment (AidanBeltonS) [reply sycl/include/CL/sycl/detail/pi.h:452]: I agree this does break the ABI. As a user visible enum is changing.
However, the current values do not conform to the SYCL2020 spec. So I believe that the enum values need to change to conform with the description of mem_advise "A value of 0 reverts the advice for ptr to the default behavior". 
Let me know if there is a better way of handling this without changing the value, many thanks?

**Comment by AidanBeltonS** [comment]
> Sorry @AidanBeltonS . I completely missed this PR. Does #6281 cover everything that this changes or should we coordinate the patches?

No worries. I think #6281 covers all everything that does not require an ABI breaking window so I don't think anything needs to be coordinated. I can update this PR later so that it just makes the enum changes.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 2]
```diff
@@ -5144,6 +5144,17 @@ pi_result cuda_piextUSMEnqueueMemAdvise(pi_queue queue, const void *ptr,
...
+    case PI_MEM_ADVISE_RESET:
+      PI_CHECK_ERROR(cuMemAdvise((CUdeviceptr)ptr, length,
+                                 CU_MEM_ADVISE_UNSET_READ_MOSTLY,
+                                 queue->get_context()->get_device()->get()));
+      PI_CHECK_ERROR(cuMemAdvise((CUdeviceptr)ptr, length,
+                                 CU_MEM_ADVISE_UNSET_PREFERRED_LOCATION,
+                                 queue->get_context()->get_device()->get()));
+      PI_CHECK_ERROR(cuMemAdvise((CUdeviceptr)ptr, length,
+                                 CU_MEM_ADVISE_UNSET_ACCESSED_BY,
+                                 queue->get_context()->get_device()->get()));
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:5156]: These two shouldn't be called unless the device has `CU_DEVICE_ATTRIBUTE_CONCURRENT_MANAGED_ACCESS`.

Comment (AidanBeltonS) [reply sycl/plugins/cuda/pi_cuda.cpp:5161]: I have updated this to check for `CU_DEVICE_ATTRIBUTE_CONCURRENT_MANAGED_ACCESS` for `PI_MEM_ADVICE_RESET`

**Review by steffenlarsen — COMMENTED** [review]
Thanks, @AidanBeltonS! Changes look good to me, but since it changes the PI API and break backwards compatability, the PI version should be incremented as detailed at the top of pi.h.

File: sycl/include/sycl/detail/pi.h [hunk 3]
```diff
@@ -49,7 +49,7 @@
 // NOTE that this results in a changed API for `piProgramGetBuildInfo`.
 
 #define _PI_H_VERSION_MAJOR 9
```
Comment (steffenlarsen) [inline sycl/include/sycl/detail/pi.h:51]: The major version should also be incremented as the change to the enum technically breaks backwards compatibility. Likewise a comment should be added to the above list of changes.

Comment (AidanBeltonS) [reply sycl/include/sycl/detail/pi.h]: Thanks for the clarification. I have incremented the major version and added a comment.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by steffenlarsen** [comment]
Test failure in Basic/image/image_write_fp16.cpp. @AidanBeltonS - Is this related to this patch?

**Comment by AidanBeltonS** [comment]
It appears to be unrelated to this patch. There is no use of mem_advice or USM within the test.

**Comment by steffenlarsen** [comment]
Do you know if it's a known failure? If not, could you please open an issue for investigating it?

**Comment by AidanBeltonS** [comment]
I do not think it is a known problem. I have created an issue to track it #6491.

**Comment by steffenlarsen** [comment]
Perfect, thank you! In that case, let's get this merged.

**Pull request merged**